### PR TITLE
Changed sympify.py to support numpy arrays

### LINKED ISSUE #10295: sympy.S handles numpy arrays incorrectly.
There are currently (at least) two problems:

First problem is implicit conversion to finite precision floating point values instead of `sympy.Integer` when applicable:

```
In [1]: import numpy, sympy

In [2]: print(numpy.__version__, sympy.__version__)
('1.10.1', '0.7.7.dev')

In [3]: sympy.S(numpy.array([1], dtype=numpy.int64))
Out[3]: 1.00000000000000
```

The other problem is that array data are interpreted as shapes somehow:

```
In [4]: out = sympy.S(numpy.array([1, 2, 3], dtype=numpy.int64))

In [5]: out.shape
Out[5]: (1, 2, 3)

In [6]: out
Out[6]: 
array([[[ 0.,  0.,  0.],
        [ 0.,  0.,  0.]]])
```

This causes major problems when doing e.g. `rref` on a numpy integer array (which we arguably should support).

### PR DESCRIPTION
**Pull request by tokencolour**
Trying to fix #10295
Currently gives output as a list.
Problem of conversion to float solved.
Still left to figure out how to type change the list
to Matrix type object.

### CODE REVIEW

**Comment by moorepants** [comment]
NumPy is an optional dependency of SymPy. You will need to ensure that something like this works fine if NumPy is not installed.

**Comment by tokencolour** [comment]
I'll make sure I'll catch the Exception `Attribute Error` and try again. 
This solution, though,  doesn't solve the purpose fully. Please see [this](https://github.com/sympy/sympy/issues/10295#issuecomment-169646474).

**Comment by Upabjojr** [comment]
SymPy master now has N-dim arrays in _sympy.tensor.array_.

I suggest a numpy array be converted into an _ImmutableDenseNDimArray_.

**Comment by tokencolour** [comment]
@Upabjojr Converting to `ImmutableDenseNDimArray` is a good idea since we'll be able to convert it into `Matrix` type by `tomatrix` attribute. `rref` and other methods in `Matrix` API might  come in handy that way.

**Comment by tokencolour** [comment]
But wouldn't that limit the use since `tomatrix` method can convert only 2-dim array? @Upabjojr , What do you think? Please have a look at the changes I've made.

File: sympy/core/sympify.py [hunk 1]
```diff
@@ -266,6 +270,23 @@ def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
...
 
+    if np is not None:
+        if isinstance(a, np.ndarray):
+            #Delayed import to avoid circular imports
+            from sympy.tensor.array.dense_ndim_array import ImmutableDenseNDimArray
+            from sympy.utilities.iterables import flatten
+            try:
+                shape = a.shape
+                if len(a.shape) == 1:
+                    shape = (1,) + shape
```
Comment (Upabjojr) [inline sympy/core/sympify.py]: why this line? Shouldn't shape remain the same?

File: sympy/core/sympify.py [hunk 2]
```diff
@@ -266,6 +270,23 @@ def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
...
+                a = flatten(a)
+                return ImmutableDenseNDimArray([sympify(x, locals=locals,
```
Comment (Upabjojr) [inline sympy/core/sympify.py]: No need to `sympify`, it is done by _ImmutableDenseNDimArray_ (though without all those parameters).

**Comment by Upabjojr** [comment]
> But wouldn't that limit the use since tomatrix method can convert only 2-dim array?

Well, N-dim arrays are a very recent addition to SymPy. The code can be changed, though I'm not sure whether _tomatrix_ should support ranks other than 2.

File: sympy/core/sympify.py (hunk 1 above)
Comment (tokencolour) [inline sympy/core/sympify.py]: I did this because of an error which kept cropping up which I am unable to reproduce now. :unamused:

**Comment by tokencolour** [comment]
>The code can be changed, though I'm not sure whether tomatrix should support ranks other than 2.
Yes. Maybe you are right. It wouldn't make much sense otherwise.

**Comment by tokencolour** [comment]
@Upabjojr , Please take a look at this. Thank you for your comments.

File: sympy/core/sympify.py [hunk 3]
```diff
@@ -266,6 +270,20 @@ def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
...
+            #Delayed import to avoid circular imports
+            from sympy.tensor.array.dense_ndim_array import ImmutableDenseNDimArray
+            from sympy.utilities.iterables import flatten
+            try:
+                shape = a.shape
+                a = flatten(a)
+                return ImmutableDenseNDimArray(a, shape)
+            except TypeError:
+                pass
+    else:
```
Comment (Upabjojr) [inline sympy/core/sympify.py]: is this _else_ necessary?

**Comment by Upabjojr** [comment]
I created a PR on your PR to add a test.

**Comment by tokencolour** [comment]
Thank you!

**Comment by Upabjojr** [comment]
Oh... test failure is a bit of a problem, need to do some changes to N-dim arrays code.

File: sympy/core/sympify.py (hunk 3 above)
Comment (tokencolour) [inline sympy/core/sympify.py]: Changed that!

**Comment by tokencolour** [comment]
I am not able to understand how these test failures are related to the changes made to sympify.py . For instance, [this](https://travis-ci.org/sympy/sympy/jobs/102161436#L3684) error ,
Wouldn't
```
assert (c1.data == c2.data)
```
be sufficient? 
@Upabjojr , Please help me understand this.

**Comment by Upabjojr** [comment]
I haven't debugged the code yet, but it looks quite clear: if a _numpy_ array algebraically interacts with a SymPy object (such as in _x*array_), the array will be sympified.

In some cases, the code is written to operate on _numpy_ objects, so they should not be converted to _SymPy_ objects.

In other cases, the operator precedence is not set, so _x*array_ doesn't call the proper array method.

**Comment by tokencolour** [comment]
@Upabjojr , Thank you for pointing the facts out otherwise I would have been lost in the numerous test failures. 
I couldn't understand functions like `__rmul__` .  How  is it different from `__mul__`? What does decorator `@call_highest_priority` do ? I looked at the code in sympy/matrices/dense.py to see how matrices are configured to combine with sympy objects. It seems like `_op_priority` is not defined for the class `ImmutableDenseNDimArray` or its ancestors per se.
Can you please help me with understanding the architecture, here. I have never really done object-oriented programming that one might call significant. However, I want to help make this class sympy compatible. This will help me learn a lot. Can you please suggest some reading material ?

**Comment by Upabjojr** [comment]
> I couldn't understand functions like __rmul__ . How is it different from __mul__?

`__rmul__` applies on the second factor if the first doesn't have `__mul__`. For example: `a*b`. If the class of `a` doesn't define `__mul__`, then Python attempts to call `b.__rmul__(a)` instead of `a.__mul__(b)`.

> It seems like _op_priority is not defined for the class ImmutableDenseNDimArray or its ancestors per se.

Yes, it should be defined in a new superclass for both `ImmutableDenseNDimArray` and `ImmutableSparseNDimArray`. Let me do it, I need to revise the class structure. N-dim arrays are a very recent addition to SymPy, so there are things to fix.

> What does decorator @call_highest_priority do ?

Makes sure that for `a*b` expressions, the method called will be the one with highest priority, that is, it will call either `a.__mul__(b)` or `b.__rmul__(a)`, not based on Python's rules, but on the value of `_op_priority`.

> However, I want to help make this class sympy compatible. This will help me learn a lot. Can you please suggest some reading material ?

I don't really know, there's the SymPy documentation. However, not all of SymPy code is well documented, and I think the best way to get into the details is by trying to contribute and then ask questions to more experienced developers.

**Comment by Upabjojr** [comment]
https://github.com/sympy/sympy/pull/10418

Here I correct the issue of `_op_priority`. Have a look at my edits, it's a good introduction to how SymPy objects should be created.

**Comment by tokencolour** [comment]
@Upabjojr , Thank you. Do keep me on track with this.

**Comment by Upabjojr** [comment]
Going on in https://github.com/sympy/sympy/pull/12507/

**Pull request closed without merging**